lib/spack/llnl/util/filesystem.py

```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
         os.unlink(path)
 
 
+class GenericVisitor(object):
+    def __init__(self, callable):
+        pass
+
+    def before_visit_dir(self, root, rel_path, depth):
+        pass
+
+    def after_visit_dir(self, root, rel_path, depth):
+        pass
+
+    def before_visit_symlinked_dir(self, root, rel_path, depth):
+        pass
+
+    def after_visit_symlinked_dir(self, root, rel_path, depth):
+        pass
+
+    def visit_file(self, root, rel_path, depth):
+        pass
+
+    def visit_symlinked_file(self, root, rel_path, depth):
+        pass
+
+
+class CallingVisitor(GenericVisitor):
```

Comment:
Firstly, I see this used in one spot: `remove_dead_links`. I don't think you actually have to traverse all symlink dirs in order to find dead links, so I'm not sure if this implementation is needed.

Secondly (and this part is only important if you disagree with the first part and think this class should remain), IMO this name doesn't explain well what this implementation does: all visitors call their supplied callback on traversed files/dirs. Probably more useful is to set the name (and docstring) to explain that this traverses all files, directories, and symlinks.